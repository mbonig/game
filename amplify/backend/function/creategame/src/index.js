const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;
const USES_SK = process.env.USES_SK;
const INDEX_NAME = process.env.INDEX_NAME || "waitingRoom";
const Crypto = require('crypto')

function randomString(size = 6) {
  return Crypto
    .randomBytes(size)
    .toString('base64')
    .slice(0, size)
    .toLowerCase();
}

function floatToTop(items, game) {
  return items.sort((x, y) => x.id === game.id ? -1 : 1);
}

exports.handler = async (event) => {
  console.log({event});

  const {myself} = event.arguments;

  if (!myself) {
    throw new Error("Please provide your name at the myself field.");
  }

  const game = {
    id: randomString(),
    players: [{name: myself}],
    status: 'Waiting',
    host: myself
  }

  // feature flag for new database table that supports sk
  if (USES_SK) {
    game.sk = game.id;
  }
  await ddb.put({TableName: TABLE, Item: game}).promise();

  const {Items: items} = await ddb.query({
    TableName: TABLE,
    IndexName: INDEX_NAME,
    KeyConditionExpression: "#status = :waiting",
    ExpressionAttributeValues: {
      ":waiting": "Waiting"
    },
    ExpressionAttributeNames: {
      "#status": "status"
    }
  }).promise();


  return floatToTop(items, game);
};
