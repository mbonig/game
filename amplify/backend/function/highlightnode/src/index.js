const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;

exports.handler = async (event) => {
    console.log({event});
    const {arguments: {id, targetNodeId, myself}} = event;
    const {Item: currentGame} = await ddb.get({Key: {id}, TableName: TABLE}).promise();
    currentGame.highlightedNodes = [(currentGame.highlightedNodes || []).filter(x=>x.username !== myself), {username: myself, targetNodeId}];
    await ddb.put({TableName: TABLE, Item: currentGame}).promise();
    return currentGame;

};
