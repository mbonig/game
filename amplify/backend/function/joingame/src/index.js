const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;
exports.handler = async (event) => {
  console.log({event, TABLE});
  const {arguments: {myself: name, id}} = event;
  const {Item: game} = await ddb.get({Key: {id: id.toLowerCase()}, TableName: TABLE}).promise();

  if (!game.players.find(p => p.name === name)) {
    game.players.push({name});
    await ddb.put({TableName: TABLE, Item: game}).promise();
  }

  return game;
};
