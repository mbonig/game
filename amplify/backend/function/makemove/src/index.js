const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;

const transportTypeTicketMap = {
  "1_slow": "slow",
  "2_medium": "medium",
  "3_fast": "fast"
}

function getTicketToUse(targetNode, currentGame, ticket) {
  const playersCurrentNode = currentGame.map.nodes.find(n => n.players.find(p => p.name === currentGame.currentTurn.name));

  const transportTypes = getAllTravel({source: playersCurrentNode, target: targetNode});
  const currentPlayer = currentGame.players.find(x => x.name === currentGame.currentTurn.name);

  if (ticket === 'black' && currentPlayer.tickets.black > 0) {
    return 'black';
  }

  for (const transportType of transportTypes) {
    const t = transportTypeTicketMap[transportType]

    if (currentPlayer.tickets[t] > 0) {
      return t;
    }
  }
  return null;
}

function checkValidMove(targetNode, currentGame, ticket) {

  if (currentGame.gameStatus && currentGame.gameStatus.status) {
    return false;
  }
  const playersCurrentNode = currentGame.map.nodes.find(n => n.players.find(p => p.name === currentGame.currentTurn.name));
  const availableTargetNodes = currentGame.map.links.filter(l => l.source === playersCurrentNode.id || l.target === playersCurrentNode.id)
    .map(l => l.source === playersCurrentNode.id ? l.target : l.source);

  const currentPlayer = currentGame.players.find(x => x.name === currentGame.currentTurn.name);
  if (ticket === 'black' && currentPlayer.tickets.black > 0) {
    return true;
  }

  if (availableTargetNodes.find(n => n === targetNode.id)) {
    // let's also check to make sure the player has a ticket to use
    const ticketUsed = getTicketToUse(targetNode, currentGame);
    if (ticketUsed === null) {
      console.log(`Didn't have a ticket to travel`);
      return false;
    }
    return true;
  }
  console.log("Couldn't find the target node in the list of availableTargetNodes", {
    targetNode,
    playersCurrentNode,
    availableTargetNodes
  });
  return false;
}

const PlayerTypes = {
  thief: "thief",
  cop: "cop"
};


function checkWinState(targetNode, currentGame) {
  if (targetNode.players.find((p) => p.type === PlayerTypes.thief)) {
    // if anyone is moving to a node where the thief is at, the game is over.
    return {
      status: 'Finished',
      winner: 'Cops'
    };
  }

  if (targetNode.players.find((p) => p.type === PlayerTypes.cop) && currentGame.currentTurn.type === PlayerTypes.thief) {
    // if the thief moves to a node that is occupied, then also game over.
    return {
      status: 'Finished',
      winner: 'Cops'
    };
  }
}

const createCurrentPlayerMapper = (player, targetNode) => (node) => {
  if (node !== targetNode) {
    node.players = node.players.filter(p => p.name !== player.name) || [];
  } else {
    node.players = [...node.players, player];
  }
  return node;
};

function getTravel(node) {
  const {source: {types: st}, target: {types: tt}} = node;
  let a = new Set(st);
  let b = new Set(tt);
  let intersection = new Set([...a].filter(x => b.has(x)));
  return [...intersection].sort().reverse()[0];
}

function getAllTravel(link) {
  const {source: {types: st}, target: {types: tt}} = link;
  let a = new Set(st);
  let b = new Set(tt);
  let intersection = new Set([...a].filter(x => b.has(x)));
  return [...intersection].sort().reverse();
}


exports.handler = async (event) => {
  console.log({event});
  const {arguments: {id, targetNodeId, myself, ticket}} = event;

  const {Item: currentGame} = await ddb.get({Key: {id}, TableName: TABLE}).promise();

  if (currentGame.currentTurn.name !== myself) {
    return currentGame;
  }
  const me = currentGame.players.find(x => x.name === myself);

  const targetNode = currentGame.map.nodes.find(x => x.id === targetNodeId);

  if (!checkValidMove(targetNode, currentGame, ticket)) {
    console.log("Not a valid move", targetNode, currentGame);
    return currentGame;
  }

  const isOver = checkWinState(targetNode, currentGame);

  const sourceNode = currentGame.map.nodes.find((n) => n.players.find((p) => p.name === currentGame.currentTurn.name));
  const moveType = getTravel({target: targetNode, source: sourceNode});
  const isThief = currentGame.currentTurn.type === PlayerTypes.thief;
  let currentIndex = currentGame.players.findIndex(x => x.name === currentGame.currentTurn.name);
  if (currentIndex >= currentGame.players.length - 1) {
    currentIndex = -1;
  }

  let ticketUsed;
  if (!ticket) {
    ticketUsed = getTicketToUse(targetNode, currentGame);
  } else {
    if (me.tickets[ticket] > 0) {
      ticketUsed = ticket;
    }
  }

  const newPlayers = currentGame.players.map(p => {
    if (p.name === currentGame.currentTurn.name) {
      p.tickets[ticketUsed] = p.tickets[ticketUsed] - 1;
    }
    return p;
  });

  const currentPlayer = currentGame.currentTurn;
  const currentMap = currentGame.map;
  const thiefMoves = isThief ? [...(currentGame.thiefMoves || []), {
    type: ticket === 'black' ? 'black' : moveType,
    nodeId: targetNode.id
  }] : currentGame.thiefMoves;
  const newGame = {
    ...currentGame,
    players: newPlayers,
    currentTurn: currentGame.players[currentIndex + 1],
    map: {
      nodes: currentMap.nodes.map(createCurrentPlayerMapper(currentPlayer, targetNode)),
      links: [...currentMap.links]
    },
    gameStatus: isOver,
    thiefMoves
  };

  await ddb.put({TableName: TABLE, Item: newGame}).promise();

  return newGame;

};
