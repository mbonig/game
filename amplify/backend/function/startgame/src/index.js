const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;

const TransportTypes = {
  slow: "1_slow",
  medium: "2_medium",
  fast: "3_fast",
};

function getRandomNode(nodes) {
  return nodes[~~(Math.random() * nodes.length)];
}

function setPlayersOnMap(map, players) {
  const newMap = {
    links: [...map.links],
    nodes: map.nodes.map(n => ({...n, players: []})),
  };

  for (const player of players) {
    let randomNode = getRandomNode(newMap.nodes);
    while (randomNode.players.length > 0) {
      randomNode = getRandomNode(newMap.nodes);
    }
    randomNode.players.push(player);

  }
  return newMap;
}

function generateMap() {
  return {
    nodes: [
      {
        id: 1,
        type: [TransportTypes.slow, TransportTypes.medium, TransportTypes.fast],
        fx: 557,
        fy: -69
      },
      {id: 2, type: [TransportTypes.slow], fx: 276, fy: -59},
      {id: 3, type: [TransportTypes.medium, TransportTypes.slow], fx: 582, fy: 214},
      {id: 4, type: [TransportTypes.fast, TransportTypes.slow], fx: 56, fy: 18},
      {
        id: 5,
        type: [TransportTypes.fast, TransportTypes.medium, TransportTypes.slow],
        players: [],
        fx: 42,
        fy: 263
      },
      {
        id: 6,
        type: [TransportTypes.medium, TransportTypes.slow],
        players: [],
        fx: -110,
        fy: 342
      },
      {
        id: 7,
        type: [TransportTypes.fast, TransportTypes.slow],
        players: [],
        fx: 276,
        fy: 203
      },
      {
        id: 8,
        type: [TransportTypes.medium, TransportTypes.slow],
        players: [],
        fx: -189,
        fy: -93
      },
      {id: 9, type: [TransportTypes.slow], players: [], fx: 98, fy: 480},
      {
        id: 10,
        type: [TransportTypes.medium, TransportTypes.slow],
        players: [],
        fx: 387,
        fy: 478
      },
    ],
    links: [
      {source: 1, target: 2},
      {source: 1, target: 3},
      {source: 2, target: 3},
      {source: 2, target: 4},
      {source: 4, target: 5},
      {source: 5, target: 6},
      {source: 3, target: 7},
      {source: 5, target: 7},
      {source: 8, target: 4},
      {source: 8, target: 6},
      {source: 9, target: 10},
      {source: 3, target: 10},
      {source: 9, target: 6},
      {source: 9, target: 5},

    ]
  };
}


exports.handler = async (event) => {
  console.log({event, TABLE});
  const {arguments: {id}} = event;
  const {Item: game} = await ddb.get({Key: {id}, TableName: TABLE}).promise();

  if (game.status !== "Waiting") {
    console.warn("Tried to start a non-waiting game...", game);
    return game;
  }

  const players = game.players.map(x => ({...x, type: 'cop'}));
  const thief = players[~~(Math.random() * players.length)];
  thief.type = "thief";
  if (game.status === 'Waiting') {
    let newGame = {
      ...game,
      players,
      map: setPlayersOnMap(generateMap(), players),
      status: 'Started',
      currentTurn: thief,
      thiefMoves: []
    };
    await ddb.put({TableName: TABLE, Item: newGame}).promise();
  }

  return game;
};
