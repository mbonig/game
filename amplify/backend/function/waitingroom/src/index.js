const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();
const TABLE = process.env.TABLE;

exports.handler = async (event) => {
  console.log({event});

  const {Items: games} = await ddb.query({
    TableName: TABLE,
    IndexName: "waitingRoom",
    KeyConditionExpression: "#status = :waiting",
    ExpressionAttributeValues: {
      ":waiting": "Waiting"
    },
    ExpressionAttributeNames: {
      "#status": "status"
    }
  }).promise();

  // we also want to get all the games the current player is already in that are still going...

  return games;
};
