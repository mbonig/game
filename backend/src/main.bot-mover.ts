import {GameState} from "../../models";

const AWS = require('aws-sdk');

// @ts-ignore
export function makeMove(game: GameState, botName: string) {
  return {targetNodeId: "", ticket: ""};
}

export const handler = async (event: any) => {
  console.log(JSON.stringify(event, null, 2));

  const {Records: records} = event;

  for (const record of records) {
    const {dynamodb: {NewImage}} = record;
    const game = AWS.DynamoDB.Converter.unmarshall(NewImage);

    if (game.id === game.sk) {
      const botName = "bot1"
      const {targetNodeId, ticket} = makeMove(game, botName);
      const timestamp = new Date().toISOString();
      const newMove = {
        id: game.id,
        sk: `MOVE#${botName}#${timestamp}`,
        player: botName,
        timestamp,
        targetNodeId,
        ticket
      }
      console.log({newMove});
    }
  }
}