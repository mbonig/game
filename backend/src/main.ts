import {App, Construct, Stack, StackProps} from '@aws-cdk/core';
import {AppSyncTransformer} from 'cdk-appsync-transformer';
import * as path from "path";
import {NodejsFunction} from "@aws-cdk/aws-lambda-nodejs";
import {AttributeType, BillingMode, ProjectionType, StreamViewType, Table} from "@aws-cdk/aws-dynamodb";
import {readdirSync} from 'fs';

interface EthrwarsStackProps extends StackProps {
  functions: string[];
}

export class EthrwarsStack extends Stack {
  // @ts-ignore
  private gameTable: Table;
  // @ts-ignore
  private transformer: AppSyncTransformer;
  // @ts-ignore
  private waitingRoomIndexName: string;

  constructor(scope: Construct, id: string, props: EthrwarsStackProps) {
    super(scope, id, props);

    // define resources here...

    this.createApi();

    this.createTable();

    for (const functionName of props.functions) {
      this.createDataSource(functionName);
    }

  }

  private createTable() {
    this.gameTable = new Table(this, 'gametable', {
      partitionKey: {
        name: "id",
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING
      },
      billingMode: BillingMode.PROVISIONED,
      readCapacity: 5,
      writeCapacity: 5,
      stream: StreamViewType.NEW_IMAGE,
    });

    this.waitingRoomIndexName = "waitingRoom";
    this.gameTable.addGlobalSecondaryIndex({
      indexName: this.waitingRoomIndexName,
      partitionKey: {
        name: 'status',
        type: AttributeType.STRING
      },
      readCapacity: 5,
      writeCapacity: 5,
      projectionType: ProjectionType.ALL
    });
  }

  private createApi() {
    this.transformer = new AppSyncTransformer(this, 'game-api', {
      schemaPath: path.join(__dirname, '../../amplify/backend/api/ethrwars/schema.graphql')
    });
  }

  private createDataSource(functionName: string) {
    const lambda = new NodejsFunction(this, functionName + '-lambda', {
      entry: path.join(__dirname, `../../amplify/backend/function/${functionName}/src/index.js`),
      environment: {
        TABLE: this.gameTable.tableName,
        INDEX_NAME: this.waitingRoomIndexName,
        USES_SK: "true"
      }
    });

    this.transformer.addLambdaDataSourceAndResolvers(functionName + '-${env}', `${functionName}_datasource`, lambda, {});

    this.gameTable.grantReadWriteData(lambda);
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const functionNames = readdirSync(path.join(__dirname, '../../amplify/backend/function'));
new EthrwarsStack(app, 'scotland-yard-api', {env: devEnv, functions: functionNames});

app.synth();