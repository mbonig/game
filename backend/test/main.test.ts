import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { EthrwarsStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new EthrwarsStack(app, 'test', {functions: []});

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});