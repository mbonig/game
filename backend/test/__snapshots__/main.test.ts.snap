// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "appsyncGraphQLEndpointOutput": Object {
      "Description": "Output for aws_appsync_graphqlEndpoint",
      "Value": Object {
        "Fn::GetAtt": Array [
          "gameapiappsyncnestedstackNestedStackappsyncnestedstackNestedStackResource83C11ED4",
          "Outputs.testgameapiappsyncnestedstackgameapiapi67279992GraphQLUrl",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241ArtifactHashE2BB9B59": Object {
      "Description": "Artifact hash for asset \\"07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241\\"",
      "Type": "String",
    },
    "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241S3BucketFE995707": Object {
      "Description": "S3 bucket for asset \\"07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241\\"",
      "Type": "String",
    },
    "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241S3VersionKey5D9ED222": Object {
      "Description": "S3 key for asset version \\"07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5ArtifactHash37ABB87F": Object {
      "Description": "Artifact hash for asset \\"b651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5\\"",
      "Type": "String",
    },
    "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5S3Bucket9F32EB9B": Object {
      "Description": "S3 bucket for asset \\"b651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5\\"",
      "Type": "String",
    },
    "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5S3VersionKey687D33BB": Object {
      "Description": "S3 key for asset version \\"b651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "botmover4AF87A38": Object {
      "DependsOn": Array [
        "botmoverServiceRoleDefaultPolicy743208CC",
        "botmoverServiceRoleCF90BC43",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5S3Bucket9F32EB9B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5S3VersionKey687D33BB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb651575e43ff24f88e40e2cefad132155ebc53515eaa4bb61c88d4f812ee5be5S3VersionKey687D33BB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "botmoverServiceRoleCF90BC43",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "botmoverDynamoDBEventSourcetestgametable8A3B9165ED52B9E4": Object {
      "Properties": Object {
        "BatchSize": 100,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "gametableAE31DE88",
            "StreamArn",
          ],
        },
        "FunctionName": Object {
          "Ref": "botmover4AF87A38",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "botmoverLogRetentionF39CA71F": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "botmover4AF87A38",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "botmoverServiceRoleCF90BC43": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "botmoverServiceRoleDefaultPolicy743208CC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "gametableAE31DE88",
                  "StreamArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "botmoverServiceRoleDefaultPolicy743208CC",
        "Roles": Array [
          Object {
            "Ref": "botmoverServiceRoleCF90BC43",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "gameapiappsyncnestedstackNestedStackappsyncnestedstackNestedStackResource83C11ED4": Object {
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241S3BucketFE995707",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241S3VersionKey5D9ED222",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters07fef7367213c5bf00c0a57bc4813db9fbd1423d9e191f69ae2f0d12e25fc241S3VersionKey5D9ED222",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
    },
    "gametableAE31DE88": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "status",
            "AttributeType": "S",
          },
        ],
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "waitingRoom",
            "KeySchema": Array [
              Object {
                "AttributeName": "status",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_IMAGE",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
